proxy/tests/transparency.rs

```diff
@@ -311,5 +319,136 @@ fn http1_get_doesnt_add_transfer_encoding() {
+    }
+}
+
+#[test]
+fn http1_one_connection_per_host() {
+    let _ = env_logger::try_init();
+
+    let srv = server::http1().route("/", "hello").run();
+    let ctrl = controller::new()
+        .run();
+    let proxy = proxy::new().controller(ctrl).inbound(srv).run();
+
+    let client = client::http1(proxy.inbound, "foo.bar");
+
+    let inbound = &proxy.inbound_server.as_ref()
+        .expect("no inbound server!");
+
+    // Make a request with the header "Host: foo.bar". After the request, the
+    // server should have seen one connection.
+    let res1 = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "foo.bar")
+    );
+    assert_eq!(res1.status(), http::StatusCode::OK);
+    assert_eq!(res1.version(), http::Version::HTTP_11);
+    assert_eq!(inbound.connections(), 1);
+
+    // Another request with the same host. The proxy may reuse the connection.
+    let res1 = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "foo.bar")
+    );
+    assert_eq!(res1.status(), http::StatusCode::OK);
+    assert_eq!(res1.version(), http::Version::HTTP_11);
+    assert_eq!(inbound.connections(), 1);
+
+    // Make a request with a different Host header. This request must use a new
+    // connection.
+    let res2 = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "bar.baz"));
+    assert_eq!(res2.status(), http::StatusCode::OK);
+    assert_eq!(res2.version(), http::Version::HTTP_11);
+    assert_eq!(inbound.connections(), 2);
+
+    let res2 = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "bar.baz"));
+    assert_eq!(res2.status(), http::StatusCode::OK);
+    assert_eq!(res2.version(), http::Version::HTTP_11);
+    assert_eq!(inbound.connections(), 2);
+
+    // Make a request with a different Host header. This request must use a new
+    // connection.
+    let res3 = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "quuuux.com"));
+    assert_eq!(res3.status(), http::StatusCode::OK);
+    assert_eq!(res3.version(), http::Version::HTTP_11);
+    assert_eq!(inbound.connections(), 3);
+}
+
+#[test]
+fn http1_requests_without_host_have_unique_connections() {
+    let _ = env_logger::try_init();
+
+    let srv = server::http1().route("/", "hello").run();
+    let ctrl = controller::new()
+        .run();
+    let proxy = proxy::new().controller(ctrl).inbound(srv).run();
+
+    let client = client::http1(proxy.inbound, "foo.bar");
+
+    let inbound = &proxy.inbound_server.as_ref()
+        .expect("no inbound server!");
+
+    // Make a request with no Host header and no authority in the request path.
+    let res = client.request(client.request_builder("/")
+        .version(http::Version::HTTP_11)
+        .header("host", "")
+    );
```

Comment:
We should also be asserting the expected values of the request URI and the `Host` header field as seen by the server in each of these test cases. That would make it clear exactly how we're modifying the request. Right now it is really unclear what the AuthorityRewriting stuff is doing. Especially, I want to make sure that in the case where we use SO_ORIGINAL_DST that we're not adding the SO_ORIGINAL_DST to the Request URI or to the Host header in the request.

Reply:
@briansmith 
> Especially, I want to make sure that in the case where we use SO_ORIGINAL_DST that we're not adding the SO_ORIGINAL_DST to the Request URI or to the Host header in the request.

I believe we _are_ currently adding the SO_ORIGINAL_DST to the request URI, but that behaviour was unchanged in this PR. I'm happy to change it, either in this PR or in a follow-up, but I think we should check with @seanmonstar first to ensure that we don't rely on this behavior elsewhere. Off the top of my head, I believe changing `reconstruct_uri` to not put the SO_ORIGINAL_DST in the request URI will require some reworking of existing telemetry code as well.

Reply:
We do modify the request `Uri` that we've accepted, however once dispatched to hyper, unless told to explicitly (via `req.set_proxy(true)`), it won't use the `absolute-form`, but just the path.

So it should be fine to check the `req.uri()` in the test server route.